riscv-runtime/src/io.rs

```diff
@@ -7,6 +9,94 @@ use powdr_riscv_syscalls::Syscall;
 use alloc::vec;
 use alloc::vec::Vec;
 
+/// An iterator over the static prover data.
+#[derive(Copy, Clone)]
+pub struct ProverDataReader {
+    remaining_data: &'static [u32],
+}
+
+use spin::Mutex;
+
+static PROVER_DATA_READER: Mutex<Option<ProverDataReader>> = Mutex::new(None);
+
+// Initialize the reader once and get a mutable reference
+fn get_prover_data_reader() -> spin::MutexGuard<'static, Option<ProverDataReader>> {
+    let mut reader = PROVER_DATA_READER.lock();
+    if reader.is_none() {
+        *reader = Some(ProverDataReader::new());
+    }
+    reader
+}
+
+impl ProverDataReader {
+    /// Creates an iterator over the static prover data.
+    ///
+    /// A newly created iterator will start at the beginning of the prover data.
+    pub fn new() -> Self {
+        extern "C" {
+            // The prover data start and end symbols. Their addresses are set by the linker.
+            static __powdr_prover_data_start: u32;
+            static __powdr_prover_data_end: u32;
+        }
+        const POWDR_PAGE_SIZE: isize = 2048;
+
+        unsafe {
+            // We skip the first page of the prover data, as it used as salt to
+            // randomize its merkle tree node.
+            let region_start: *const u32 = &__powdr_prover_data_start;
+            let data_start = region_start.byte_offset(POWDR_PAGE_SIZE);
+            let data_end: *const u32 = &__powdr_prover_data_end;
+
+            let prover_data_section =
+                slice::from_raw_parts(data_start, data_end.offset_from(data_start) as usize);
+
+            // The first word of the prover data section is the total number of words the user wrote.
+            let (&total_words, remaining_data) = prover_data_section.split_first().unwrap();
+
+            let remaining_data = &remaining_data[..total_words as usize];
+            Self { remaining_data }
+        }
+    }
+}
+
+impl Iterator for ProverDataReader {
+    type Item = &'static [u8];
+
+    /// Returns the next slice of prover data.
+    ///
+    /// Because it is in static memory, the reference can be stored, passed around and will always be valid.
+    ///
+    /// The start of the slice is guaranteed to be 4-bytes aligned.
+    fn next(&mut self) -> Option<&'static [u8]> {
+        if self.remaining_data.is_empty() {
+            return None;
+        }
+
+        let (&len_bytes, remaining) = self.remaining_data.split_first().unwrap();
+        let len_words = (len_bytes + 3) / 4;
+        let (data, remaining) = remaining.split_at(len_words as usize);
+        self.remaining_data = remaining;
+
+        // SAFETY: It is safe to cast an u32 slice to an u8 slice, which is the most general type.
+        unsafe {
+            let data_ptr = data.as_ptr() as *const u8;
+            Some(slice::from_raw_parts(data_ptr, len_bytes as usize))
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
```

Comment:
It should be cheap to turn this into an exact size iterator (either iterate over the sizes when creating the iterator, or write the object count as the second word (right after the total number of words). Maybe its worth it.

Reply:
but then, we don't really expose this iterator to the user, only the read function, so I think it doesn't matter

Reply:
We don't? We do.

`ProverDataReader` is public and can be instantiated whenever the user wants. The stateful global instance is just to mimic read from stdin and allegedly be more user friendly.

That said, I am not sure complicating even more the binary format of the user data to make it an exact size iterator is worth it...